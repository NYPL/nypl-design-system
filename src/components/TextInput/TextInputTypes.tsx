// HTML Input types as defined by MDN: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input
export enum TextInputTypes {
  email = "email",
  hidden = "hidden",
  // eslint-disable-next-line id-blacklist
  number = "number",
  password = "password",
  text = "text",
  // textarea = "textarea",
  tel = "tel",
  url = "url",
}

export enum TextInputFormats {
  email = "jdoe@domain.com",
  hidden = "",
  // eslint-disable-next-line id-blacklist
  number = "",
  password = "",
  text = "",
  tel = "(123) 123-1234",
  textarea = "",
  url = "https://domain.com",
}
/*
export enum TextInputPatterns {
  // eslint-disable-next-line prettier/prettier
  email = "[a-z0-9._%+-]+@[a-z0-9.-]+.[a-z]{2,}$",
  hidden = "",
  // eslint-disable-next-line id-blacklist
  number = "",
  password = "",
  text = "",
  // eslint-disable-next-line prettier/prettier
  tel = "([0-9]{3}) [0-9]{3}-[0-9]{4}",
  textarea = "textarea",
  url = "https?://.+",
}
*/
