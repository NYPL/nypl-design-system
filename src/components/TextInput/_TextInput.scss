$break-input-mobile: 400px;

.textinput {
  @include space-stack-xs;

  .inputfield {
    @include placeholder {
      color: var(--ui-gray-dark);
      font-style: italic;
      line-height: 21px;
    }

    border: 1px solid var(--ui-gray-medium);
    padding: var(--space-xs) var(--space-s);
    width: 100%;
    // Include this once textarea is included as an input "type" option.
    // &--textarea {
    //   min-height: var(--space-xxl);
    // }
    &:hover {
      border-color: var(--ui-gray-dark);
    }

    &:disabled {
      background-color: var(--ui-gray-xxlight);
      color: var(--ui-gray-xdark);
      pointer-events: none;
    }

    &:active,
    &:focus,
    &:active:hover,
    &:focus:hover {
      border: 1px solid var(--ui-focus);
      box-shadow: 0;
      outline: 1px solid var(--ui-focus);
      z-index: 9999;
    }

    &--error {
      border: 1px solid var(--ui-error);

      &:hover,
      &:active,
      &:focus,
      &:active:hover,
      &:focus:hover {
        border: 1px solid var(--ui-error);
      }
    }
  }

  > label {
    font-weight: 300;
  }

  > .helper-text {
    margin-bottom: 0;
    margin-top: var(--space-xs);
  }
}
