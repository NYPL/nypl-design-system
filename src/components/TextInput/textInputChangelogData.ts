/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "Prerelease",
    version: "Prerelease",
    type: "Update",
    affects: ["Styles"],
    notes: [
      "Reduced the spacing between the field label and the field itself.",
    ],
  },
  {
    date: "2024-03-14",
    version: "3.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Chakra 2.8 update."],
  },
  {
    date: "2023-11-09",
    version: "2.1.2",
    type: "Update",
    affects: ["Accessibility"],
    notes: [
      'Added the `autoComplete` prop for setting the "autocomplete" attribute manually.',
    ],
  },
  {
    date: "2023-10-18",
    version: "2.1.0",
    type: "Bug Fix",
    affects: ["Accessibility"],
    notes: [
      "Updated so the `aria-describedby` value is not overwritten as 'undefined' when `TextInput` is part of the `DatePicker` component.",
    ],
  },
  {
    date: "2023-9-28",
    version: "2.0.0",
    type: "Update",
    affects: ["Styles"],
    notes: ["Applied Typo2023 styles, including font size and font color."],
  },
];
