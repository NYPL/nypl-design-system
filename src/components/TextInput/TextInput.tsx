import * as React from "react";
import bem from "../../utils/bem";
import { TextInputTypes, TextInputFormats } from "./TextInputTypes";
import Label from "../Label/Label";
import HelperErrorText from "../HelperErrorText/HelperErrorText";
import generateUUID from "../../helpers/generateUUID";

export interface InputProps {
  /** Additional attributes to pass to the `<input>` tag */
  attributes?: { [key: string]: any };
  /** HTML Input types as defined by MDN: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input */
  type?: TextInputTypes;
  /** Will add 'aria-required: true' to input */
  required?: boolean;
  /** Provides text for a `Label` component if `showLabel` is set to true; populates a `aria-label` sttribute if `showLabel` is set to false. */
  labelText: string;
  /** Offers the ability to show the label onscreen or hide it. Refer to the `labelText` property for more information. */
  showLabel?: boolean;
  /** Populates the HelperErrorText for standard state */
  helperText?: string;
  /** Populates the HelperErrorText for error state */
  errorText?: string;
  /** Populates the placeholder of the input */
  placeholder?: string;
  /** Populates the value of the input */
  value?: string;
  /** className you can add in addition to 'textinput' */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Helper for modifiers array; adds 'errored' styling */
  errored?: boolean;
  /** Adds the 'disabled' prop to the input when true */
  disabled?: boolean;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** The action to perform on the `<input>`'s onChange function  */
  onChange?: (
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>
  ) => void;
}

const TextInput = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref?) => {
    const {
      type = TextInputTypes.text,
      required,
      labelText,
      showLabel = true,
      helperText,
      errorText,
      placeholder,
      value,
      className,
      disabled,
      errored,
      onChange,
    } = props;

    const attributes = props.attributes || {};
    const modifiers = props.modifiers ? props.modifiers : [];
    const id = props.id || generateUUID();

    if (!showLabel) attributes["aria-label"] = labelText;
    if (helperText) attributes["aria-describedby"] = helperText;

    if (errored) {
      modifiers.push("error");
    }
    // WILL COME BACK TO OFFERING TEXTAREA AS AN OPTION
    // if (type === TextInputTypes.textarea) {
    //   modifiers.push("textarea");
    // }

    const errorOutput = errorText
      ? errorText
      : "There is an error related to this field.";

    let footnote;
    if (errored) {
      footnote = errorOutput;
    } else {
      footnote = helperText;
    }

    if (
      type === TextInputTypes.tel ||
      type === TextInputTypes.url ||
      type === TextInputTypes.email
    ) {
      const example = TextInputFormats[type];
      footnote = (
        <>
          Ex: {example}
          <br />
          {footnote}
        </>
      );
    }

    const textField = (
      <input
        id={id}
        className={bem("inputfield", modifiers, "", [className])}
        type={type}
        value={value}
        aria-required={required}
        aria-hidden={type === TextInputTypes.hidden}
        disabled={disabled}
        required={required}
        placeholder={placeholder}
        onChange={onChange}
        ref={ref}
        {...attributes}
      />
    );
    // const textareaField = (
    //   <textarea
    //     id={id}
    //     className={bem("inputfield", modifiers, "", [className])}
    //     aria-required={required}
    //     aria-hidden={type === TextInputTypes.hidden}
    //     disabled={disabled}
    //     required={required}
    //     placeholder={placeholder}
    //     onChange={onChange}
    //     // ref={ref}
    //     {...attributes}
    //   >
    //     {value}
    //   </textarea>
    // );
    // WILL COME BACK TO OFFERING TEXTAREA AS AN OPTION
    // const fieldOutput =
    //   type === TextInputTypes.textarea ? textareaField : textField;
    const fieldOutput = textField;

    const optReqFlag = required ? "Required" : "Optional";
    const transformedInput = (
      <div className="textinput">
        {labelText && showLabel && type !== TextInputTypes.hidden && (
          <Label htmlFor={id} optReqFlag={optReqFlag} id={id + `-label`}>
            {labelText}
          </Label>
        )}
        {fieldOutput}
        {((helperText && type !== TextInputTypes.hidden) || errored) && (
          <HelperErrorText isError={errored} id={id + `-helperText`}>
            {footnote}
          </HelperErrorText>
        )}
      </div>
    );

    return transformedInput;
  }
);

TextInput.displayName = "TextInput";

export default TextInput;
