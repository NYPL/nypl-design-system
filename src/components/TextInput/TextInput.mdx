import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as TextInputStories from "./TextInput.stories";
import ComponentChangelogTable from "../../utils/ComponentChangelogTable";
import Link from "../Link/Link";
import { changelogData } from "./textInputChangelogData";

<Meta of={TextInputStories} />

# TextInput

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.22.0`   |
| Latest            | `2.1.0`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#labelling-variations" target="_self">Labelling Variations</Link>}
- {<Link href="#browser-states" target="_self">Browser States</Link>}
- {<Link href="#isclearable-button" target="_self">isClearable Button</Link>}
- {<Link href="#number-type" target="_self">Number Type</Link>}
- {<Link href="#html-in-helper-text" target="_self">HTML in Helper Text</Link>}
- {<Link href="#textarea" target="_self">Textarea</Link>}
- {<Link href="#changelog" target="_self">Changelog</Link>}

## Overview

<Description of={TextInputStories} />

This component renders a text input form element. See below for configuration
information. For optimal accessibility, the `labelText` property is a required
prop, regardless of the label visibility. Additionally, while the `id` prop is
optional, a unique `id` attribute is necessary for accessibility. If the prop
is left blank, a value will be generated for you.

## Component Props

<Canvas of={TextInputStories.WithControls} />

<ArgTypes of={TextInputStories.WithControls} />

## Accessibility

The `TextInput` component renders an HTML `<input>` element which is accessible
via keyboard and screen reader. The `aria-required` and `required` attributes are
set with the `isRequired` prop. The `aria-disabled` and `disabled` attributes are
set with the `isDisabled` prop. The `aria-invalid` is set with the `isInvalid`
prop.

Internally, a `Label` is associated with the `<input>` element. When `showLabel`
is set to false, the `<input>` element's `aria-label` attribute is set to the
required `labelText` value.

The `helperText` and the `invalidText` are associated with the `<input>` element
through the `aria-describedby` attribute.

When the `isClearable` prop is set to `true`, as text is added to the input
field a close button is rendered to clear the input value. Once the close button
is clicked, the input value will be cleared, the close button will become hidden,
and focus will be set to the input element.

When the `type` prop is set to `"textarea"`, the `<textarea>` element
is rendered instead of the `<input>` element. This element follows all the same
accessibility rules described above.

If the `type` prop is set to `"email"`, `"tel"`, or `"url"`, the appropriate
`autocomplete` attribute will be added. The autocomplete attribute makes inputs
easier and more efficient to complete for all users.

Resources:

- [MDN input: The Input (Form Input) element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input)
- [MDN textarea: The Textarea element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea)
- [MDN HTML attribute: Autocomplete](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)
- [Chakra UI Input](https://chakra-ui.com/docs/components/form/input)
- [Chakra UI Textarea](https://chakra-ui.com/docs/components/form/textarea)

## Labelling Variations

A TextInput can be rendered with or without visible labels. When `showLabel` is
set to `false`, an `aria-label` attribute is added to the input/textarea element
to maintain accessibility. If the component needs to be required, the
`showRequiredLabel` prop can be used to show or hide the "(Required)" text
within the `label` element.

<Canvas of={TextInputStories.LabellingVariations} />

## Browser States

<Canvas of={TextInputStories.BrowserStates} />

## isClearable Button

Set the `isClearable` prop to `true` to render a button that clears the input
value. This close button will render when the input has a value and disappear
when the input is empty, including after it is clicked.

This button has visually hidden text, a "close" icon, and renders on top
of the input field on the right side. This clear button does not render when
the `type` prop is set to `"textarea"`.

Note: This works best when the `TextInput` is used as an uncontrolled component.

<Canvas of={TextInputStories.isClearableButton} />

### Controlled isClearableCallback Example

A `TextInput` component is controlled when the `value` and `onChange` props are
set. For a controlled component, the `isClearableCallback` prop can be used to
clear the input value in the consuming application. This callback is only called
when the close button is clicked.

<Source
  code={`
// ExampleControlled.js
export const App = () => {
  const [value, setValue] = React.useState("");
  return (
    <TextInput
      id="isClearable-controlled-example"
      isClearable
      isClearableCallback={() => setValue("")}
      labelText="What is your favorite color?"
      onChange={(e) => setValue(e.target.value)}
      placeholder="i.e. blue, green, etc."
      value={value}
    />
  );
};
`}
  language="jsx"
/>

<Canvas of={TextInputStories.ControlledExample} />

## Number Type

The `TextInput` component can be configured to render a number input by setting
`type="number"`. In this input type, the `maxLength` prop does not restrict the
length of the number that is passed to the input. This is the expected behavior
as the native HTML input and `maxlength` attribute behave this way.

For this type, the `min` and `max` props do restrict the value of the number
input that is passed. This can be seen by using the up and down arrows to
increase or decrease the number value, respectively. However, if the `max` prop
is set to "200", for example, the user can still enter a number such as "400".
The consuming application must handle validating the entered number value.

If the `min` prop value is greater than the `max` prop value, it will trigger
the invalid state.

<Canvas of={TextInputStories.NumberType} />

### Alternate Number Pattern

Another pattern for entering numbers in a restrictive way is to use the
`maxLength` and `pattern` props. Together, these props can be set to define how
many digits and what characters are allowed. For example, if a zip code is
required, the `maxLength` prop can be set to "5" and the `pattern` prop can be
set to `"[0-9]\*"` to allow only numbers to be entered and limit the number of
characters to 5.

Note that the `type` prop is not set to "number" in this example and that the
consuming application must handle validating the value entered. Browsers will
also provide their own validation for the input.

<Source
  code={`
<TextInput
  helperText="Enter a 5-digit zip code."
  id="number-pattern"
  labelText="Enter a zip code"
  maxLength={5}
  pattern="[0-9]*"
  placeholder="i.e. 10018"
/>
`}
  language="jsx"
/>

<Canvas of={TextInputStories.AlternateNumberPattern} />

## HTML in Helper Text

HTML can be passed into the `helperText` prop as a string or HTML.

<Source
  code={`
helperText="Choose <b>wisely!</b>"
// or
helperText={<>Choose <b>wisely!</b></>}
`}
  language="jsx"
/>

<Canvas of={TextInputStories.HTMLHelperText} />

## Textarea

The TextInput component includes a multiline `textarea` form field. To render a
`textarea` element, pass `"textarea"` for the `type` prop. All the variations
described above are available for the `textarea` option.

<Canvas of={TextInputStories.Textarea} />

## Changelog

<ComponentChangelogTable changelogData={changelogData} />
