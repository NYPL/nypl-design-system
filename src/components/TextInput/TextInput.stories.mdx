import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";

import TextInput from "./TextInput";
import { TextInputTypes } from "./TextInputTypes";
import { action } from "@storybook/addon-actions";
import { getCategory } from "../../utils/componentCategories";

<Meta
  title={getCategory("TextInput")}
  component={TextInput}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=11895%3A547",
    },
  }}
/>

# TextInput

<Description of={TextInput} />

This component renders a text input form element. See below for configuration information. For optimal accessibility, the `labelText` property is a required prop, regardless of the label visibility. Additionally, while the `id` prop is optional, a unique `id` attribute is necessary for accessibility. If the prop is left blank, a value will be generated for you.

<Preview withToolbar>
  <Story
    name="TextInput"
    args={{
      id: "myTextInput",
      type: "text",
      labelText: "What is your favorite color?",
      showLabel: true,
      placeholder: "e.g. blue, green, etc.",
      helperText: "Choose wisely.",
      errorText: "Waaaaaaaa!",
      errored: false,
      required: true,
      disabled: false,
      attributes: {
        onFocus: action("focus"),
        onBlur: action("blur"),
        onChange: action("onChange"),
        maxLength: 10,
        tabIndex: 0,
        autoFocus: true,
        pattern: "[a-z0-9]",
      },
    }}
  >
    {args => <TextInput {...args} />}
  </Story>
</Preview>

<ArgsTable story="TextInput" />

## Labelling Variations

A TextInput can be rendered with or without visible labels.

<Preview withToolbar>
  <Story
    name="Labelling Variations"
    args={{
      labelText: "What is your favorite color?",
      showLabel: false,
    }}
  >
    {args => <TextInput {...args} />}
  </Story>
</Preview>

<Canvas>
  <TextInput labelText="What is your favorite color?" showLabel={true} />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={false}
    placeholder="i.e. blue, green, etc."
  />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={true}
    placeholder="i.e. blue, green, etc."
  />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={false}
    placeholder="i.e. blue, green, etc."
    helperText="Choose wisely!"
  />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={true}
    placeholder="i.e. blue, green, etc."
    helperText="Choose wisely!"
  />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={false}
    placeholder="i.e. blue, green, etc."
    errorText="Waaaaaa!"
    errored={true}
  />
</Canvas>

<Canvas>
  <TextInput
    labelText="What is your favorite color?"
    showLabel={true}
    placeholder="i.e. blue, green, etc."
    errorText="Waaaaaa!"
    errored={true}
  />
</Canvas>

<!-- Include this once textarea is included as an input "type" option.
## Textarea

The TextInput component includes a multiline `textarea` form field.

<Preview withToolbar>
  <Story
    name="Textarea"
    args={{
      type: "textarea",
      labelText: "In less than 500 words, describe your favorite colors?",
      showLabel: true,
      placeholder: "Essay question...",
      helperText: "Let it all out.",
    }}
  >
    {args => <TextInput {...args} />}
  </Story>
</Preview>
-->
