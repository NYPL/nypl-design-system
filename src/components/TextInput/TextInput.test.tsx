import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";
import { stub, spy } from "sinon";

import HelperErrorText from "../HelperErrorText/HelperErrorText";
import TextInput from "./TextInput";
import { TextInputTypes } from "./TextInputTypes";
import generateUUID from "../../helpers/generateUUID";

describe("Renders TextInput with visible label", () => {
  let container;
  let changeHandler;
  let focusHandler;

  before(() => {
    focusHandler = stub();
    changeHandler = stub();
    container = Enzyme.mount(
      <TextInput
        id={"myTextInput"}
        labelText={"Custom Input Label"}
        required={true}
        placeholder={"Input Placeholder"}
        type={TextInputTypes.text}
        attributes={{ onFocus: focusHandler }}
        onChange={changeHandler}
      ></TextInput>
    );
  });

  it("Renders Input", () => {
    expect(container.find("input").exists()).to.equal(true);
  });

  it("Renders as type `text`", () => {
    expect(container.find("input").prop("type")).to.equal("text");
  });

  it("Renders label with label text", () => {
    expect(container.find("label").exists()).to.equal(true);
  });

  it("Renders correct label text", () => {
    expect(container.find("label").text()).to.contain("Custom Input Label");
  });

  it("Renders label's `for` attribute pointing at ID from input", () => {
    expect(container.find("label").prop("htmlFor")).to.equal("myTextInput");
  });

  it("Renders placeholder text", () => {
    expect(container.find("input").props().placeholder).to.equal(
      "Input Placeholder"
    );
  });

  it("Adds aria-required prop if input is required", () => {
    expect(container.find("input").prop("aria-required")).to.equal(true);
  });

  it("Allows user to pass in additional attributes", () => {
    container.find("input").simulate("focus");
    expect(focusHandler.callCount).to.equal(1);
  });

  it("Changing the value calls the onChange handler", () => {
    container.find("input").simulate("change", { target: { value: "Hello" } });

    expect(changeHandler.callCount).to.equal(1);
  });
});

describe("Renders TextInput with auto-generated ID, hidden label and visible helper text", () => {
  let container;
  let generateUUIDSpy;

  before(() => {
    generateUUIDSpy = spy(generateUUID);
    container = Enzyme.mount(
      <TextInput
        labelText={"Custom Input Label"}
        showLabel={false}
        helperText={"Custom Helper Text"}
        required={true}
        placeholder={"Input Placeholder"}
        type={TextInputTypes.text}
      ></TextInput>
    );
  });

  it("Renders Input component", () => {
    expect(container.find("input").exists()).to.equal(true);
  });

  it("Calls a UUID generation method if no ID is passed as a prop", () => {
    expect(container.find("input").props()).to.have.property("id");
    expect(generateUUIDSpy.called);
  });

  it("Does not renders Label component", () => {
    expect(container.find("label").exists()).to.equal(false);
  });

  it("Renders custom aria-label", () => {
    expect(container.find("input").prop("aria-label")).to.equal(
      "Custom Input Label"
    );
  });

  it("Renders HelperErrorText component", () => {
    expect(container.find(".helper-text").exists()).to.equal(true);
  });

  it("Renders aria-describedby with helper text", () => {
    expect(container.find("input").prop("aria-describedby")).to.equal(
      "Custom Helper Text"
    );
  });
});

describe("TextInput shows error state", () => {
  let container;
  before(() => {
    container = Enzyme.mount(
      <TextInput
        id={"myTextInputError"}
        labelText={"Custom Input Label"}
        helperText={"Custom Helper Text"}
        errorText={"Custom Error Text"}
        placeholder={"Input Placeholder"}
        errored={true}
        type={TextInputTypes.text}
      ></TextInput>
    );
  });

  it("Renders Input component", () => {
    expect(container.find("input").exists()).to.equal(true);
  });

  it("Renders Label component", () => {
    expect(container.find("label").exists()).to.equal(true);
  });

  it("Renders HelperErrorText component", () => {
    expect(container.find(".helper-text").exists()).to.equal(true);
  });

  it("Input shows error state", () => {
    expect(container.find("input").hasClass("inputfield--error")).to.equal(
      true
    );
  });

  it("Helper text shows error state", () => {
    expect(
      container.find(".helper-text").hasClass("helper-text--error")
    ).to.equal(true);
  });

  it("Helper text shows error text", () => {
    expect(container.find(".helper-text").text()).to.equal("Custom Error Text");
  });
});

describe("TextInput Group", () => {
  let container;
  before(() => {
    container = Enzyme.mount(
      <fieldset>
        <legend>Input Group Label</legend>

        <TextInput
          id="input1"
          required={true}
          type={TextInputTypes.text}
          labelText="For"
          helperText="Input 1 Helper Text"
        />

        <TextInput
          id="input2"
          required={true}
          type={TextInputTypes.text}
          labelText="To"
          helperText="Input 2 Helper Text"
        />

        <HelperErrorText isError={true} id={"sharedHelperText"}>
          Additional Error Text
        </HelperErrorText>
      </fieldset>
    );
  });

  it("Renders fieldset", () => {
    expect(container.find("fieldset").exists()).to.equal(true);
  });

  it("Renders legend", () => {
    expect(container.find("legend").exists()).to.equal(true);
  });

  it("Renders two inputs", () => {
    expect(container.find("input")).to.have.length(2);
  });
});

describe("Renders HTML attributes passed through the `attributes` prop", () => {
  const onChangeSpy = stub();
  const onBlurSpy = stub();
  let container;
  before(() => {
    container = Enzyme.mount(
      <TextInput
        id="inputID-attributes"
        labelText="Input Label"
        placeholder={"Input Placeholder"}
        type={TextInputTypes.text}
        attributes={{
          onChange: onChangeSpy,
          onBlur: onBlurSpy,
          maxLength: 10,
          tabIndex: 0,
        }}
      ></TextInput>
    );
  });

  it("Has a maxlength for the input element", () => {
    expect(container.find("input").prop("maxLength")).to.equal(10);
  });

  it("Has a tabIndex", () => {
    expect(container.find("input").prop("tabIndex")).to.equal(0);
  });

  it("Calls the onChange function", () => {
    expect(onChangeSpy.callCount).to.equal(0);
    container.find("input").simulate("change");
    expect(onChangeSpy.callCount).to.equal(1);
  });

  it("Calls the onBlur function", () => {
    expect(onBlurSpy.callCount).to.equal(0);
    container.find("input").simulate("blur");
    expect(onBlurSpy.callCount).to.equal(1);
  });
});

describe("Forwarding refs", () => {
  it("Passes the ref to the input element", () => {
    const ref = React.createRef<HTMLInputElement>();
    const container = Enzyme.mount(
      <TextInput
        id="inputID-attributes"
        labelText="Input Label"
        placeholder={"Input Placeholder"}
        type={TextInputTypes.text}
        ref={ref}
      ></TextInput>
    );
    expect(container.find("input").instance()).to.equal(ref.current);
  });
});

describe("Hidden input", () => {
  it("Renders a hidden type input", () => {
    const container = Enzyme.mount(
      <TextInput
        id="inputID-hidden"
        labelText={"Hidden Input Label"}
        type={TextInputTypes.hidden}
        value="hidden"
      />
    );

    const input = container.find("input");

    expect(input.prop("aria-hidden")).to.equal(true);
    expect(input.prop("value")).to.equal("hidden");
  });
});
