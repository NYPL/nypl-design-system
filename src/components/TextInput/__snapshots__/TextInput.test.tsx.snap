// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Snapshots renders the text input UI snapshot correctly 1`] = `
<div
  className="css-1u8qly9"
  id="myTextarea-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextarea"
    id="myTextarea-label"
  >
    Custom textarea Label
  </label>
  <textarea
    autocomplete={null}
    className="chakra-textarea css-0"
    disabled={false}
    id="myTextarea"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Textarea Placeholder"
    required={false}
    step={null}
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 2`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
    <span>
       (Required)
    </span>
  </label>
  <input
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 3`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
    <span>
       (Obiligatoire)
    </span>
  </label>
  <input
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 4`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
  </label>
  <input
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={false}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 5`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <input
    aria-label="Custom Input Label"
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 6`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
    <span>
       (Required)
    </span>
  </label>
  <input
    aria-describedby="myTextInput-helperText"
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Custom helper text",
      }
    }
    data-isinvalid={false}
    id="myTextInput-helperText"
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 7`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
    <span>
       (Required)
    </span>
  </label>
  <input
    aria-describedby="myTextInput-helperText"
    aria-invalid={true}
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
  <div
    aria-atomic={true}
    aria-live="polite"
    className="css-1xdhyk6"
    dangerouslySetInnerHTML={
      Object {
        "__html": "There is an error related to this field.",
      }
    }
    data-isinvalid={true}
    id="myTextInput-helperText"
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 8`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
    <span>
       (Required)
    </span>
  </label>
  <input
    aria-required={true}
    autocomplete={null}
    className="chakra-input css-0"
    disabled={true}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={true}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 9`] = `
<div
  className="css-1u8qly9"
  id="myTextInput-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="myTextInput"
    id="myTextInput-label"
  >
    Custom Input Label
  </label>
  <input
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="myTextInput"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={false}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 10`] = `
<div
  className="css-1y4kn3e"
  id="chakra-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="chakra"
    id="chakra-label"
  >
    Custom Input Label
  </label>
  <input
    autocomplete={null}
    className="chakra-input css-0"
    disabled={false}
    id="chakra"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={false}
    step={null}
    type="text"
    value=""
  />
</div>
`;

exports[`UI Snapshots renders the text input UI snapshot correctly 11`] = `
<div
  className="css-1u8qly9"
  data-testid="props"
  id="props-wrapper"
>
  <label
    className="css-1xdhyk6"
    htmlFor="props"
    id="props-label"
  >
    Custom Input Label
  </label>
  <input
    autocomplete={null}
    className="chakra-input css-0"
    data-testid="props"
    disabled={false}
    id="props"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Input Placeholder"
    required={false}
    step={null}
    type="text"
    value=""
  />
</div>
`;
